File: server/src/constants/judgeColors.js
/**
 * Shared judge/tablet color palette. Keys are stable; use same in admin and Flutter app.
 * Each entry: { key, hex }
 */
const JUDGE_COLORS = [
  { key: 'red', hex: '#dc3545' },
  { key: 'blue', hex: '#0d6efd' },
  { key: 'green', hex: '#198754' },
  { key: 'orange', hex: '#fd7e14' },
  { key: 'yellow', hex: '#ffc107' },
  { key: 'purple', hex: '#6f42c1' },
  { key: 'pink', hex: '#d63384' },
  { key: 'teal', hex: '#20c997' },
  { key: 'cyan', hex: '#0dcaf0' },
  { key: 'lime', hex: '#84cc16' },
  { key: 'indigo', hex: '#6610f2' },
  { key: 'brown', hex: '#795548' },
  { key: 'gold', hex: '#ffb300' },
  { key: 'magenta', hex: '#c2185b' },
  { key: 'dark_blue', hex: '#1e3a5f' },
];

const COLOR_BY_KEY = Object.fromEntries(JUDGE_COLORS.map((c) => [c.key, c.hex]));

function getHex(key) {
  if (!key || typeof key !== 'string') return null;
  return COLOR_BY_KEY[key.trim().toLowerCase()] || null;
}

function isValidKey(key) {
  return !!getHex(key);
}

/** V0/admin sends hex; map to backend key. Exact match or first key. */
function hexToKey(hex) {
  if (!hex || typeof hex !== 'string') return '';
  const h = hex.trim().toLowerCase();
  if (!h.startsWith('#')) return '';
  for (const { key, hex: hx } of JUDGE_COLORS) {
    if (hx && hx.toLowerCase() === h) return key;
  }
  return JUDGE_COLORS[0] ? JUDGE_COLORS[0].key : '';
}

module.exports = {
  JUDGE_COLORS,
  COLOR_BY_KEY,
  getHex,
  isValidKey,
  hexToKey,
};

File: tablet_app/lib/config/judge_colors.dart
import 'package:flutter/material.dart';

/// Shared judge/tablet color palette. Keys must match backend (server/src/constants/judgeColors.js).
class JudgeColors {
  JudgeColors._();

  static const List<JudgeColorEntry> palette = [
    JudgeColorEntry('red', Color(0xFFDC3545)),
    JudgeColorEntry('blue', Color(0xFF0D6EFD)),
    JudgeColorEntry('green', Color(0xFF198754)),
    JudgeColorEntry('orange', Color(0xFFFD7E14)),
    JudgeColorEntry('yellow', Color(0xFFFFC107)),
    JudgeColorEntry('purple', Color(0xFF6F42C1)),
    JudgeColorEntry('pink', Color(0xFFD63384)),
    JudgeColorEntry('teal', Color(0xFF20C997)),
    JudgeColorEntry('cyan', Color(0xFF0DCAF0)),
    JudgeColorEntry('lime', Color(0xFF84CC16)),
    JudgeColorEntry('indigo', Color(0xFF6610F2)),
    JudgeColorEntry('brown', Color(0xFF795548)),
    JudgeColorEntry('gold', Color(0xFFFFB300)),
    JudgeColorEntry('magenta', Color(0xFFC2185B)),
    JudgeColorEntry('dark_blue', Color(0xFF1E3A5F)),
  ];

  static Color? colorForKey(String? key) {
    if (key == null || key.isEmpty) return null;
    final k = key.trim().toLowerCase();
    // Palette key lookup (e.g. 'red', 'blue')
    for (final e in palette) {
      if (e.key == k) return e.color;
    }
    // Hex fallback (e.g. '#dc3545') — in case server sends hex instead of key
    if (k.startsWith('#') && (k.length == 7 || k.length == 9)) {
      try {
        final hexStr = k.length == 7 ? 'ff${k.substring(1)}' : k.substring(1);
        final colorInt = int.parse(hexStr, radix: 16);
        return Color(colorInt);
      } catch (_) {}
    }
    return null;
  }

  static String? keyForColor(Color color) {
    final hex = '#${color.value.toRadixString(16).padLeft(8, '0').substring(2)}';
    for (final e in palette) {
      final eHex = '#${e.color.value.toRadixString(16).padLeft(8, '0').substring(2)}';
      if (eHex.toUpperCase() == hex.toUpperCase()) return e.key;
    }
    return null;
  }

  /// Hex string for injection / display (e.g. #198754).
  static String? hexForKey(String? key) {
    final c = colorForKey(key);
    if (c == null) return null;
    return '#${c.red.toRadixString(16).padLeft(2, '0')}${c.green.toRadixString(16).padLeft(2, '0')}${c.blue.toRadixString(16).padLeft(2, '0')}';
  }

  static bool isValidKey(String? key) => colorForKey(key) != null;
}

class JudgeColorEntry {
  final String key;
  final Color color;

  const JudgeColorEntry(this.key, this.color);
}
